# Notebook 5 – Aggregation Features

Aggregation features in machine learning summarize raw, detailed event data over specific groups or time windows into meaningful statistical signals for a model

## Why Aggregation Features Are Useful
Raw transaction data has one row per purchase — but ML models often need to understand behavior at a **higher level** (per customer, per product, or per order). Aggregation features summarize many rows into a single meaningful number (like total spend or average order size), which helps models capture patterns like customer value, buying habits, and product popularity that a single transaction row can't show.


In [1]:
import pandas as pd
df = pd.read_csv('data.csv', encoding='ISO-8859-1')
df['TotalPrice'] = df['Quantity'] * df['UnitPrice']
df[['InvoiceNo', 'CustomerID', 'Quantity', 'UnitPrice', 'TotalPrice']].head()

,InvoiceNo,CustomerID,Quantity,UnitPrice,TotalPrice
0,536365,17850.0,6,2.55,15.30
1,536365,17850.0,6,3.39,20.34
2,536365,17850.0,8,2.75,22.00
3,536365,17850.0,6,3.39,20.34
4,536365,17850.0,6,3.39,20.34


## 1. Count

**Explanation:** Counts how many transactions/rows exist for a group, e.g. per customer.

**Example:** Customer 17850 appears in 20 rows → Count = 20

**Why we use / need:** Shows how active or engaged a customer or product is — more transactions can mean higher engagement or popularity.

**When we use:** Use when measuring activity level, e.g. number of purchases, number of reviews.

In [2]:
customer_count = df.groupby('CustomerID')['InvoiceNo'].count().reset_index()
customer_count.columns = ['CustomerID', 'PurchaseCount']
customer_count.head()

,CustomerID,PurchaseCount
0,12346.0,2
1,12347.0,182
2,12348.0,31
3,12349.0,73
4,12350.0,17


**Code Explanation:** .groupby('CustomerID') groups rows by customer; .count() counts non-null entries in InvoiceNo per group.

## 2. Sum

**Explanation:** Adds up a numeric column for each group, e.g. total money spent by a customer.

**Example:** Customer 17850 spent 15.30 + 20.34 + ... → Sum = Total Spending

**Why we use / need:** Total spend is one of the strongest indicators of customer value — key for CLV and VIP segmentation.

**When to use:** Use for total spending, total quantity sold, or total revenue per product/customer.

In [3]:
customer_spend = df.groupby('CustomerID')['TotalPrice'].sum().reset_index()
customer_spend.columns = ['CustomerID', 'TotalSpending']
customer_spend.head()

,CustomerID,TotalSpending
0,12346.0,0.00
1,12347.0,4310.00
2,12348.0,1797.24
3,12349.0,1757.55
4,12350.0,334.40


**Code Explanation:** .sum() adds up TotalPrice for all rows belonging to the same CustomerID.

## 3. Mean

**Explanation:** Calculates the average value of a numeric column per group, e.g. average order value.

**Example:** Customer spent 15.30, 20.34, 10.00 → Mean = 15.21

**Why we use / need:** Average order value shows typical spending behavior, independent of how many times they bought — useful for comparing customer types.

**When to use:** Use for average order value, average product price, or average rating.

In [4]:
customer_avg = df.groupby('CustomerID')['TotalPrice'].mean().reset_index()
customer_avg.columns = ['CustomerID', 'AvgOrderValue']
customer_avg.head()

,CustomerID,AvgOrderValue
0,12346.0,0.000000
1,12347.0,23.681319
2,12348.0,57.975484
3,12349.0,24.076027
4,12350.0,19.670588


**Code Explanation:** `.mean()` computes the average `TotalPrice` per customer group

## 4. Median

**Explanation:** Finds the middle value of a numeric column per group, less affected by extreme values than mean.

**Example:** Order values [10, 12, 500] → Median = 12 (Mean would be skewed by 500)

**Why we use / need:** Useful when data has outliers (e.g. one huge bulk order) — median gives a more "typical" picture than mean.

**When to use:** Use when spending or quantity data is skewed by a few very large transactions.

In [5]:
customer_median = df.groupby('CustomerID')['TotalPrice'].median().reset_index()
customer_median.columns = ['CustomerID', 'MedianOrderValue']
customer_median.head()

,CustomerID,MedianOrderValue
0,12346.0,0.00
1,12347.0,17.00
2,12348.0,41.76
3,12349.0,17.70
4,12350.0,19.80


**Code Explanation:** `.median()` returns the middle value of `TotalPrice` for each customer group

## 5. Minimum

**Explanation:** Finds the smallest value of a numeric column per group.

**Example:** Customer's orders [15.30, 20.34, 5.00] → Min = 5.00

**Why we use / need:** Helps identify the smallest purchase a customer made — useful to detect low-value or trial purchases.

**When to use:** Use when you need the lowest transaction amount, lowest price, or earliest small order.

In [6]:
customer_min = df.groupby('CustomerID')['TotalPrice'].min().reset_index()
customer_min.columns = ['CustomerID', 'MinOrderValue']
customer_min.head()

,CustomerID,MinOrderValue
0,12346.0,-77183.60
1,12347.0,5.04
2,12348.0,13.20
3,12349.0,6.64
4,12350.0,8.50


**Code Explanation:** .min() returns the smallest TotalPrice value within each customer group.

## 6. Maximum

**Explanation:** Finds the largest value of a numeric column per group.

**Example:** Customer's orders [15.30, 20.34, 500.00] → Max = 500.00

**Why we use / need:** Identifies the biggest single purchase — useful for spotting high-value transactions or VIP behavior.

**When to use:** Use when you need the largest transaction amount, e.g. maximum transaction amount per customer.

In [7]:
customer_max = df.groupby('CustomerID')['TotalPrice'].max().reset_index()
customer_max.columns = ['CustomerID', 'MaxOrderValue']
customer_max.head()

,CustomerID,MaxOrderValue
0,12346.0,77183.6
1,12347.0,249.6
2,12348.0,240.0
3,12349.0,300.0
4,12350.0,40.0


**Code Explanation:** .max() returns the largest TotalPrice value within each customer group.

## 7. Standard Deviation

**Explanation:** Measures how much a customer's spending varies — low value means consistent spending, high value means unpredictable spending.

**Example:** Orders [15, 16, 14] → low Std Dev (consistent); Orders [5, 200, 10] → high Std Dev (inconsistent)

**Why we use / need:** Helps distinguish steady, predictable customers from erratic, unpredictable ones — useful for risk scoring or personalized offers.

**When to use:** Use when analyzing spending consistency or variability in behavior.

In [8]:
customer_std = df.groupby('CustomerID')['TotalPrice'].std().reset_index()
customer_std.columns = ['CustomerID', 'StdOrderValue']
customer_std.head()

,CustomerID,StdOrderValue
0,12346.0,109154.093913
1,12347.0,23.289902
2,12348.0,48.514857
3,12349.0,34.655913
4,12350.0,7.275538


**Code Explanation:** .std() calculates the standard deviation of TotalPrice per customer group.

## 8. Group-Based Aggregation (Multiple Stats at Once)

**Explanation:** Instead of calculating each statistic separately, `.agg()` computes multiple aggregations together in one step.

**Example:** For each customer → Count, Sum, Mean, Min, Max of TotalPrice all in one table

**Why we use / need:** Saves time and keeps related features organized together, especially when building a feature table for modeling.

**When to use:** Use when you need several summary statistics for the same group at once, e.g. building a customer feature table.

In [9]:
customer_agg = df.groupby('CustomerID')['TotalPrice'].agg(['count', 'sum', 'mean', 'min', 'max', 'std']).reset_index()
customer_agg.columns = ['CustomerID', 'PurchaseCount', 'TotalSpending', 'AvgOrderValue', 'MinOrderValue', 'MaxOrderValue', 'StdOrderValue']
customer_agg.head()

,CustomerID,PurchaseCount,TotalSpending,AvgOrderValue,MinOrderValue,MaxOrderValue,StdOrderValue
0,12346.0,2,0.00,0.000000,-77183.60,77183.6,109154.093913
1,12347.0,182,4310.00,23.681319,5.04,249.6,23.289902
2,12348.0,31,1797.24,57.975484,13.20,240.0,48.514857
3,12349.0,73,1757.55,24.076027,6.64,300.0,34.655913
4,12350.0,17,334.40,19.670588,8.50,40.0,7.275538


**Code Explanation:** .agg([...]) applies a list of aggregation functions at once to the grouped column, producing one row per customer with multiple stats.

## 9. Customer-Level Aggregation

**Explanation:** Combines multiple columns' aggregations into a single feature table describing each customer's overall behavior.

**Example:** Customer 17850 → 20 purchases, $ 450 total spend, $ 22.50 avg order

**Why we use / need:** Customer-level features are the foundation of customer segmentation, churn prediction, and lifetime value models — raw transaction rows alone can't be fed into these models directly.

**When to use:** Use whenever the ML task is about predicting or segmenting customers (not individual transactions).

In [10]:
customer_features = df.groupby('CustomerID').agg(
    PurchaseCount=('InvoiceNo', 'count'),
    TotalSpending=('TotalPrice', 'sum'),
    AvgOrderValue=('TotalPrice', 'mean'),
    UniqueProducts=('StockCode', 'nunique')
).reset_index()
customer_features.head()

,CustomerID,PurchaseCount,TotalSpending,AvgOrderValue,UniqueProducts
0,12346.0,2,0.00,0.000000,1
1,12347.0,182,4310.00,23.681319,103
2,12348.0,31,1797.24,57.975484,22
3,12349.0,73,1757.55,24.076027,73
4,12350.0,17,334.40,19.670588,17


**Code Explanation:** .agg(name=('column', 'function')) creates named aggregated columns in one step; nunique() counts distinct products bought by each customer.

## 10. Product-Level Aggregation

**Explanation:** Summarizes sales behavior for each product instead of each customer.

**Example:** Product 85123A → sold 500 times, total revenue $1275

**Why we use / need:** Helps identify best-selling products, pricing patterns, and demand — useful for inventory planning and recommendation systems.

**When to use:** Use when the ML task involves product performance, demand forecasting, or recommendation engines.

In [11]:
product_features = df.groupby('StockCode').agg(
    TimesSold=('InvoiceNo', 'count'),
    TotalQuantitySold=('Quantity', 'sum'),
    TotalRevenue=('TotalPrice', 'sum'),
    AvgUnitPrice=('UnitPrice', 'mean')
).reset_index()
product_features.head()

,StockCode,TimesSold,TotalQuantitySold,TotalRevenue,AvgUnitPrice
0,10002,73,1037,759.89,1.056849
1,10080,24,495,119.09,0.376667
2,10120,30,193,40.53,0.210000
3,10123C,4,-13,3.25,0.487500
4,10123G,1,-38,0.00,0.000000


**Code Explanation:** Groups data by StockCode (product) and calculates count, total quantity, total revenue, and average price per product.

## 11. Transaction-Level Aggregation

**Explanation:** Summarizes each individual order (invoice) rather than customer or product — e.g. how many items and total value per order.

**Example:** Invoice 536365 → 5 items, total value $65.30

**Why we use / need:** Useful for understanding basket size and order value patterns, which matter for pricing strategy and fraud detection (e.g. unusually large orders).

**When to use:** Use when the ML task is about analyzing or predicting order-level behavior, like basket size or order value.

In [12]:
transaction_features = df.groupby('InvoiceNo').agg(
    ItemsInOrder=('StockCode', 'count'),
    OrderTotalValue=('TotalPrice', 'sum')
).reset_index()
transaction_features.head()

,InvoiceNo,ItemsInOrder,OrderTotalValue
0,536365,7,139.12
1,536366,2,22.20
2,536367,12,278.73
3,536368,4,70.05
4,536369,1,17.85


**Code Explanation:** Groups data by InvoiceNo (each order) and calculates number of items and total order value.